File: 2360-longest-cycle-in-a-graph/2360-longest-cycle-in-a-graph.cpp
class Solution {
public:
    int longestCycle(vector<int>& edges) {
        int n = edges.size();
        vector<int> g[n];
        for (int i = 0; i < n; ++i) {
            if (edges[i] != -1) {
                g[i].push_back(edges[i]);
            }
        }
        int longestCycleLength = 0;
        vector<bool> visited(n, false), callStack(n, false);
        vector<int> parent(n, -1);
        function<void(int, int)> dfs = [&](int u, int p) {
            visited[u] = true;
            callStack[u] = true;
            parent[u] = p;
            for (auto &v : g[u]) {
                if (!visited[v]) {
                    dfs(v, u);
                } else {
                    if (callStack[v] == true) {
                        int cycleLength = 1;
                        int node = u;
                        while (node != v) {
                            cycleLength++;
                            node = parent[node];
                        }
                        longestCycleLength = max(longestCycleLength, cycleLength);
                    }
                }
            }
            callStack[u] = false;
        };
        for (int i = 0; i < n; ++i) {
            if (!visited[i]) {
                dfs(i, i);
            }
        }
        return (longestCycleLength == 0) ? -1 : longestCycleLength;
    }
};

File: 2360-longest-cycle-in-a-graph/2360-longest-cycle-in-a-graph.java
class Solution {
    public int longestCycle(int[] edges) {
        int n = edges.length;
        ArrayList<ArrayList<Integer>> g = new ArrayList<>();
        for (int i = 0; i < n; ++i) {
            g.add(new ArrayList<>());
        }
        for (int i = 0; i < n; ++i) {
            if (edges[i] != -1) {
                g.get(i).add(edges[i]);
            }
        }
        int longestCycle = -1;
        boolean[] vis = new boolean[n];
        boolean[] rec = new boolean[n];
        int[] par = new int[n];
        Arrays.fill(par, -1);
        for (int i = 0; i < n; ++i) {
            if (!vis[i]) {
                int cycleLength = dfs(g, i, i, par, vis, rec);
                longestCycle = Math.max(longestCycle, cycleLength);
            }
        }
        return longestCycle;
    }

    public int dfs(ArrayList<ArrayList<Integer>> g, int u, int p, int[] par, boolean[] vis, boolean[] rec) {
        int currCycleLength = -1;
        vis[u] = true;
        rec[u] = true;
        par[u] = p;
        for (int v : g.get(u)) {
            if (!vis[v]) {
                int childCycleLength = dfs(g, v, u, par, vis, rec);
                currCycleLength = Math.max(currCycleLength, childCycleLength);
            } else {
                if (rec[v]) {
                    int newCycleLength = 1;
                    int currNode = u;
                    while (v != currNode) {
                        currNode = par[currNode];
                        ++newCycleLength;
                    }
                    currCycleLength = Math.max(currCycleLength, newCycleLength);
                }
            }
        }
        rec[u] = false;
        return currCycleLength;
    }
}